Finish the drake_copyable porting of framework

jwnimmer-tri opened the pull request:
These were the few cases that were non-boilerplate changes.

This is a portion of #4861.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4939)
<!-- Reviewable:end -->

jwnimmer-tri commented:
+@david-german-tri and +@sherm1 for review, please.  Please take your time on this one, the lines are relatively few but the ideas are important.

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbY4ZFtzq5N-qJ3TVkd:b-22q0xx)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I like the idea of using final to provide slice-proof copy/move functionality. But the proposed approach appears to require each final derived class to implement the base class copy/move functionality, which is brittle -- a base class should take care of its own data members. Why not do that with protected copy/move functionality in the base? 

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-Kbk86EUnYSKxad-Omiz:b-vv0nft) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
> ```objc
>   SimpleCarConfig(const SimpleCarConfig& other)
>       : systems::BasicVector<T>(K::kNumCoordinates) {
>     this->SetFromVector(other.CopyToVector());
> ```

Jeremy, this seems brittle to me. Why does the derived class have to know so much about the base class in order to copy it? I would have given BasicVector a protected copy constructor so that it could be in charge of copying its own fields. Why not do that here?

---

*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-Kbk8kz_VelRqyvOwXxZ:b1gnq0p) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
> ```objc
>  public:
>   // BasicVector cannot be copied or moved; use Clone instead.
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(BasicVector)
> ```

If the goal is for the most-derived classes to have public copy or move functionality, I would have expected BasicVector to provide protected copy/move helpers so that (potentially many) derived classes don't each have to know how to copy the base class data members. Why is this a better approach? 

---

*[drake/systems/framework/value.h, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk9c-5QZ5IiDZg9Hi2:-Kbk9c-5QZ5IiDZg9Hi3:bfabjvu) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/value.h#L32)):*
> ```objc
> class AbstractValue {
>  public:
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(AbstractValue)
> ```

One of the most basic operations to do with an AbstractValue during simulation is to swap a new one with an old one. These can be very large objects (like a delay table) and swapping between cache and an abstract state variable avoids per-step heap thrashing. Why do you want to remove that capability? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbkAZt9T7cxRtGuBp9E:bcuwk6a)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/value.h, line 32 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk9c-5QZ5IiDZg9Hi2:-KbkCd-5wL__B3Rwmi6D:bd8b4h0) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/value.h#L32)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

One of the most basic operations to do with an AbstractValue during simulation is to swap a new one with an old one. These can be very large objects (like a delay table) and swapping between cache and an abstract state variable avoids per-step heap thrashing. Why do you want to remove that capability? 
</blockquote></details>

Sorry -- withdrawn. Abstract values swaps are done by swapping the AbstractValue objects (i.e. the pointers) not their contents. Never mind!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I agree with that sentiment broadly speaking (of using `= default` copy and copy-assign), but in this case I think the PR is better without taking that approach.  Details below.

---

Review status: 6 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-KbkTR7pyMBnyQjoa_id:b-6sua1s) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Jeremy, this seems brittle to me. Why does the derived class have to know so much about the base class in order to copy it? I would have given BasicVector a protected copy constructor so that it could be in charge of copying its own fields. Why not do that here?
</blockquote></details>

I've taken a second look here and realized there was clearer implementation.  PTAL.

To my read, the subclass here now knows exactly as much as it should about the base class -- that to copy a Config the new Config has copy of the elements.  That doesn't seem brittle?  I don't (yet?) see a way for an future implementor to add a bug to this?  There is (rightly) no default ctor to BasicVector, so removing the initializer list by accident would not compile.

---

*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-KbkTraMIfL7Lf7uxkZk:b-9avu1s) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

If the goal is for the most-derived classes to have public copy or move functionality, I would have expected BasicVector to provide protected copy/move helpers so that (potentially many) derived classes don't each have to know how to copy the base class data members. Why is this a better approach? 
</blockquote></details>

I've updated the comment.  Even with `protected` access, I think a `BasicVector-` copy or copy-assign still exposes us to the slicing problem, as well as the oops-the-size-changed problem.  In subclasses that offer appropriate guarantees, they can implement what they need via the BasicVector public APIs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbkVCS_B456ctoB2p9D:b-51d2qm)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> seems right to me, thanks for doing this!

---

Reviewed 6 of 9 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 49 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-KblQDfpt9gGS8CUnzr5:-KblQDfqKhoeqym84aCB:bse4to9) ([raw file](https://github.com/robotlocomotion/drake/blob/ab42b9fee29fe8da5e1fe40c997d2897d45112f2/drake/automotive/gen/simple_car_config.h#L49)):*
> ```objc
>   /// Copy-assignment operator.
>   SimpleCarConfig& operator=(const SimpleCarConfig& other) {
>     this->SetFromVector(other.CopyToVector());
> ```

BTW, I think the `set_value(get_value())` spelling would be a bit less scary here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KblQRwTQKEccQ04KTwZ:b7k1ibn)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 7 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 49 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-KblQDfpt9gGS8CUnzr5:-KblWN6r1BhaiS4J1Sal:b-rkzrr1) ([raw file](https://github.com/robotlocomotion/drake/blob/ab42b9fee29fe8da5e1fe40c997d2897d45112f2/drake/automotive/gen/simple_car_config.h#L49)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW, I think the `set_value(get_value())` spelling would be a bit less scary here.
</blockquote></details>

Done.  (Oops.  Somehow I didn't find those methods when trying to find the best set/get pair to call.  Definitely better.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-Kbm2P7zQQAtajlaY96y:b-6zug83) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I've taken a second look here and realized there was clearer implementation.  PTAL.

To my read, the subclass here now knows exactly as much as it should about the base class -- that to copy a Config the new Config has copy of the elements.  That doesn't seem brittle?  I don't (yet?) see a way for an future implementor to add a bug to this?  There is (rightly) no default ctor to BasicVector, so removing the initializer list by accident would not compile.
</blockquote></details>

This still seems to me to be dependent on knowledge that the base class has no meaningful data content other than the value (which I admit is all that there is at the moment!). But if some future programmer were to revise BasicVector at some point, say to add a name, or provenance information, or a counter, or whatever, the copy constructor for `SimpleCarConfig`, which alleges to be a kind of BasicVector, would extract and copy only the data and nothing else. That would be a defect, but how is the future programmer (or reviewer) supposed to know that? One possible fix would be to change the documentation for BasicVector to prohibit any future additions to its data members, with an explicit promise that derived classes can count on that. I would be OK with that but it seems weird to me. 

---

*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-Kbm90vq3MNAI3q3kt5p:b-foax13) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I've updated the comment.  Even with `protected` access, I think a `BasicVector-` copy or copy-assign still exposes us to the slicing problem, as well as the oops-the-size-changed problem.  In subclasses that offer appropriate guarantees, they can implement what they need via the BasicVector public APIs.
</blockquote></details>

I think what you have now is an almost-guaranteed future slicing problem since there is nothing to stop BasicVector from growing some new data fields without the derived classes failing to copy them. Protected methods that refuse to change size would be slice-proof, since the derived classes would have only themselves to worry about. Am I misunderstanding your point?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+(status: curate commits before merging) because the intermediate commits don't compile anymore.

+(status: do not merge) while we iterate on the design.

---

Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-KbmF16k1bNPW7r--oXF:b360k2b) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This still seems to me to be dependent on knowledge that the base class has no meaningful data content other than the value (which I admit is all that there is at the moment!). But if some future programmer were to revise BasicVector at some point, say to add a name, or provenance information, or a counter, or whatever, the copy constructor for `SimpleCarConfig`, which alleges to be a kind of BasicVector, would extract and copy only the data and nothing else. That would be a defect, but how is the future programmer (or reviewer) supposed to know that? One possible fix would be to change the documentation for BasicVector to prohibit any future additions to its data members, with an explicit promise that derived classes can count on that. I would be OK with that but it seems weird to me. 
</blockquote></details>

I've added BasicVector class methods to place an abstraction barrier around this concept.  If the approach seems good, I will add unit tests for them also.

In doing this, I realized that the codegen'd `DoClone` was also defective with respect to this discussion, so I've fixed that also.

---

*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-KbmCq5pYal3R2D6tZDB:b5hofvz) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I think what you have now is an almost-guaranteed future slicing problem since there is nothing to stop BasicVector from growing some new data fields without the derived classes failing to copy them. Protected methods that refuse to change size would be slice-proof, since the derived classes would have only themselves to worry about. Am I misunderstanding your point?
</blockquote></details>

The point was that people would accidentally invoke `BasicVector<double>::operator=(const BasicVector<double>& other)` when the `this` receiver was one subclass and the `other` argument was another -- it would compile, and look totally normal at the call site, but it would be incorrect.  Making the method protected only limits the number of places the error could occur; the slicing (copying or assigning via the base class function, when subclasses have additional details) could still happen.  I believe the extra tag type in the current revision makes it unlikely to be used accidentally now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbmKkSPIIbBhYf3FgTE:b-e2l0o2)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/>

---

Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-Kbp3Z2w17GrFDv6hGQq:b4xm580) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
> One possible fix would be to change the documentation for BasicVector to prohibit any future additions to its data members, with an explicit promise that derived classes can count on that. I would be OK with that but it seems weird to me.

I'm in favor of this.  The whole point of BasicVector is that it's semantics-free.

> I've added BasicVector class methods to place an abstraction barrier around this concept. If the approach seems good, I will add unit tests for them also.

I think the cure, while clever, is worse than the disease.  `DoClone` and `operator=` were previously legible, and are now magical blobs everyone has to copy-paste.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-Kbp5lZwBSSuw1DneOOV:bkw1etm)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-Kbp6eaKuliksVZy4IWd:b-lt6dpm) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

> One possible fix would be to change the documentation for BasicVector to prohibit any future additions to its data members, with an explicit promise that derived classes can count on that. I would be OK with that but it seems weird to me.

I'm in favor of this.  The whole point of BasicVector is that it's semantics-free.

> I've added BasicVector class methods to place an abstraction barrier around this concept. If the approach seems good, I will add unit tests for them also.

I think the cure, while clever, is worse than the disease.  `DoClone` and `operator=` were previously legible, and are now magical blobs everyone has to copy-paste.
</blockquote></details>

Fine by me.  @sherm1 do you agree (with documenting that there is no member data other than `value_`)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-Kbp_icdIZd9By7Q4upn:b-jzusyf) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The point was that people would accidentally invoke `BasicVector<double>::operator=(const BasicVector<double>& other)` when the `this` receiver was one subclass and the `other` argument was another -- it would compile, and look totally normal at the call site, but it would be incorrect.  Making the method protected only limits the number of places the error could occur; the slicing (copying or assigning via the base class function, when subclasses have additional details) could still happen.  I believe the extra tag type in the current revision makes it unlikely to be used accidentally now.
</blockquote></details>

Let me see if I get it now: Let classes A and B be derived from BasicVector (BV). Public methods `A::operator=(const A&)` and `B::operator=(const B&)` exist, but 
`BV::operator=(const BV&)` is protected. A public invocation of `A::operator=(const B&)` won't compile because the BV copy assignment is protected. So you are concerned that a method implemented in A, say `A::method()` will internally invoke an assignment like `A::operator=(const B& b)`? Then that will match the protected method `BV::operator=(BV(b))` and thus slice?

I'm having trouble imagining the circumstance under which one BV-derived class internally does a copy assignment to another BV-derived class. Is that a reasonable concern?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-Kbpci23qZp3FcaHHcDn:bh7y26i) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Let me see if I get it now: Let classes A and B be derived from BasicVector (BV). Public methods `A::operator=(const A&)` and `B::operator=(const B&)` exist, but 
`BV::operator=(const BV&)` is protected. A public invocation of `A::operator=(const B&)` won't compile because the BV copy assignment is protected. So you are concerned that a method implemented in A, say `A::method()` will internally invoke an assignment like `A::operator=(const B& b)`? Then that will match the protected method `BV::operator=(BV(b))` and thus slice?

I'm having trouble imagining the circumstance under which one BV-derived class internally does a copy assignment to another BV-derived class. Is that a reasonable concern?

</blockquote></details>

Just one derived class `A` is enough.  The hazard would be that code within `A` invokes `BV::operator=(const BV&)` when it meant to call `A::operator=(const A&)`.  The types being copied to and from are as desired (both are `A` at runtime), but the `operator=(const BV&)` might not copy everything.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-KbpgtKd5dLXXiKltFZd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Fine by me.  @sherm1 do you agree (with documenting that there is no member data other than `value_`)?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-Kbph_3heDAXcj497BnE:b9byjet) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Just one derived class `A` is enough.  The hazard would be that code within `A` invokes `BV::operator=(const BV&)` when it meant to call `A::operator=(const A&)`.  The types being copied to and from are as desired (both are `A` at runtime), but the `operator=(const BV&)` might not copy everything.
</blockquote></details>

I don't get how that could happen. If I write
```c++
A a1, a2;
a1 = a2;
```
how would I inadvertently invoke `BV::op=(const BV&)`?

It is certainly true that the implementation of `A::op=(const A&)` would invoke protected `BV::op=(const BV&)` explicitly, which would be the intended use, but how does it happen by accident?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-KbqoyCuLZUSN56dyiFC:b-whme3u) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
> @sherm1 do you agree (with documenting that there is no member data other than value_)?

That's acceptable but I still think it is unnecessary. PTAL at my other comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+(status: do not review) 
I am going to break out the uncontroversial pieces of this into separate PRs.

---

Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-Kbr6i6-mnFo8mRyfsJb:b-w5rnyl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
The separate PRs have been cherry-picked into this PR for now to get CI coverage of the final result.  I'm still keeping the do-no-review tag until the other PRs merge.  (At least, don't review anything other than the "Finish the drake_copyable porting of framework" commit.)

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbrQ7G4oHb4Qd48vefM:b2j9xg9)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 22 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/gen/simple_car_config.h, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk86EUnYSKxad-Omiy:-KbrQf3QorRC8Cz5KlqP:b-g1otp7) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/automotive/gen/simple_car_config.h#L46)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

> @sherm1 do you agree (with documenting that there is no member data other than value_)?

That's acceptable but I still think it is unnecessary. PTAL at my other comment.
</blockquote></details>

I have withdrawn the change that generated this discussion.

---

*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-KbrQtS4gKRvxXbG3nG-:b-ibqen4) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I don't get how that could happen. If I write
```c++
A a1, a2;
a1 = a2;
```
how would I inadvertently invoke `BV::op=(const BV&)`?

It is certainly true that the implementation of `A::op=(const A&)` would invoke protected `BV::op=(const BV&)` explicitly, which would be the intended use, but how does it happen by accident?
</blockquote></details>

So with VectorBase copy-assign hypothetically still being deleted, but BasicVector copy-assign being protected, it ends up being a little difficult to tickle the error, but it would be something like:
```c++
class MyA final : public BasicVector {
 public:
  MyA Clamp(const MyA& other) {
    auto result = this->Clone();
    if (result->NormInf() > other.NormInf()) {
      *result = other;   // Bad: slicing calls BV::op=.
    }
    return MyA(static_cast<MyA&>(*result));
  }

 private:
  // ... assume there's supplemental member data here ...
};
```

This is a bit unconvincing because a reasonable person would do `MyA result;` instead of the `Clone`, but it's still code that reads correctly when skimmed, but is subtly wrong.  Hopefully the gist is clear in that if anywhere in `MyA` ever uses `BasicVector` for a local variable type (whether explicitly, or via auto) then that code is as risk of calling `BV::op=` accidentally.  The defect surface is even wider if VectorBase turns its special member functions protected instead of deleted (as it likely should if we decide that BasicVector would have them be defaulted and protected?).

In any case, the final subclass with assignment has been removed from this PR, so I believe the discussion is moot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/systems/framework/basic_vector.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-Kbk8kz_VelRqyvOwXxY:-KbugFv8NnYMswil2Hvf:bgm34ee) ([raw file](https://github.com/robotlocomotion/drake/blob/42db44d7da7aa33f231f7fa55188a478bac087e0/drake/systems/framework/basic_vector.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

So with VectorBase copy-assign hypothetically still being deleted, but BasicVector copy-assign being protected, it ends up being a little difficult to tickle the error, but it would be something like:
```c++
class MyA final : public BasicVector {
 public:
  MyA Clamp(const MyA& other) {
    auto result = this->Clone();
    if (result->NormInf() > other.NormInf()) {
      *result = other;   // Bad: slicing calls BV::op=.
    }
    return MyA(static_cast<MyA&>(*result));
  }

 private:
  // ... assume there's supplemental member data here ...
};
```

This is a bit unconvincing because a reasonable person would do `MyA result;` instead of the `Clone`, but it's still code that reads correctly when skimmed, but is subtly wrong.  Hopefully the gist is clear in that if anywhere in `MyA` ever uses `BasicVector` for a local variable type (whether explicitly, or via auto) then that code is as risk of calling `BV::op=` accidentally.  The defect surface is even wider if VectorBase turns its special member functions protected instead of deleted (as it likely should if we decide that BasicVector would have them be defaulted and protected?).

In any case, the final subclass with assignment has been removed from this PR, so I believe the discussion is moot.
</blockquote></details>

That's a good example, but I would attribute the failure to a bad implementation of `Clone()`. Programmers will assume (reasonably) that `A->Clone()` returns an `A`. That is one of the many benefits of NVI -- `BV::Clone()` is non-virtual, `DoClone()` is virtual and covariant, `A::Clone()` can shadow `BV::Clone()` and return `unique_ptr<A>`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
PTAL.  All of the prerequisite PRs have now merged, and I think I have addressed all of the discussions on this PR. 
-(status: do not merge) -(status: do not review)  -(status: curate commits before merging) 

---

Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbujYa8QWCIWjAEbrSk:blp4x0p)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 3 of 15 files at r4.
Review status: 3 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/tools/lcm_vector_gen.py, line 91 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-KbumsCeQV2DGKVXHWdh:-KbumsCfFOoNPqaBHWLU:b-8boi5e) ([raw file](https://github.com/robotlocomotion/drake/blob/c60d81c01a249e92cadc1cc1f270c2f1f27da1bc/drake/tools/lcm_vector_gen.py#L91)):*
> ```Python
>   %(camel)s<T>* DoClone() const override {
>     auto result = new %(camel)s;
>     result->set_value(this->get_value());
> ```

Perhaps I've lost the thread now, but isn't this exactly the slicing problem we had before? It looks like the generated derived class Clone() attempts to copy the base class by using its secret knowledge that there is only one field there. I would have done this as `return new Derived(*this);` where the Derived copy constructor makes use of the protected BasicVector copy constructor. Did you redefine BasicVector to guarantee that there will never be another data member?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/tools/lcm_vector_gen.py, line 91 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-KbumsCeQV2DGKVXHWdh:-Kbup5gOMKfk9hiBiakJ:bi65jfe) ([raw file](https://github.com/robotlocomotion/drake/blob/c60d81c01a249e92cadc1cc1f270c2f1f27da1bc/drake/tools/lcm_vector_gen.py#L91)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Perhaps I've lost the thread now, but isn't this exactly the slicing problem we had before? It looks like the generated derived class Clone() attempts to copy the base class by using its secret knowledge that there is only one field there. I would have done this as `return new Derived(*this);` where the Derived copy constructor makes use of the protected BasicVector copy constructor. Did you redefine BasicVector to guarantee that there will never be another data member?
</blockquote></details>

This change only respells the logic using less-scary accessors.  I will happily revert this part of the PR if you prefer, but David had earlier suggested making this change.  It is currently correct, and any framework brittleness I will leave to framework authors to fix in follow-up PRs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 3 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/tools/lcm_vector_gen.py, line 91 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-KbumsCeQV2DGKVXHWdh:-KbuvaROg3W46RidX6bG:b69mrna) ([raw file](https://github.com/robotlocomotion/drake/blob/c60d81c01a249e92cadc1cc1f270c2f1f27da1bc/drake/tools/lcm_vector_gen.py#L91)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This change only respells the logic using less-scary accessors.  I will happily revert this part of the PR if you prefer, but David had earlier suggested making this change.  It is currently correct, and any framework brittleness I will leave to framework authors to fix in follow-up PRs.
</blockquote></details>

OK, how about adding a developer's comment on BasicVector documenting that it has just a single data member and that no more may be added without re-thinking how derived classes (including generated ones) are copied.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/tools/lcm_vector_gen.py, line 91 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-KbumsCeQV2DGKVXHWdh:-Kbv0d2IeZTS4pjbyy3O:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c60d81c01a249e92cadc1cc1f270c2f1f27da1bc/drake/tools/lcm_vector_gen.py#L91)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

OK, how about adding a developer's comment on BasicVector documenting that it has just a single data member and that no more may be added without re-thinking how derived classes (including generated ones) are copied.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 15 files at r4, 1 of 1 files at r5.
Review status: 6 of 12 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbvERY5B55al-JhOC9y:bnfp4nl)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 15 files at r4.
Review status: 6 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4939#-:-KbwDwqPh8FTmeF0GcZc:bnfp4nl)*
<!-- Sent from Reviewable.io -->
